fix typing definitions and make pycountry.db.Database generic

terencehonles opened the pull request:
There are a number of problems with pycountry's new typings and this should fix a number of them. The database objects returned "themself" when they really should have been returning the type of data they would return. I attempted using the `__class_getitem__` hook, but it doesn't look like mypy supports this. Since the `data_class` `str` attribute was only used in a couple cases I tried shifting the factory creation up, but it looks like mypy also does not like `type(name, (parents, ...), {})` in a factory or bare, so since I could _also declare the classes pretty easily with `class Name(Data): pass` I did so.

Because all the data classes now were mostly visible (with the exception of `Country` and `Subdivision`) it seemed to make sense to re-export those names as top-level variables so it makes it easier for a caller to define:

```
from pycountry import countries, Country

def filter_countries(...) -> list[Country]:
  return [c for c in countries if ...]
```

terencehonles commented on src/pycountry/__init__.py:
```diff
@@ -3,10 +3,12 @@
 
 import pycountry.db
 
+from pycountry.db import Country as Country, Subdivision as Subdivision
```
Allow these to be accessed from a base export like the other classes defined in this module

terencehonles commented on src/pycountry/__init__.py:
```diff
@@ -113,7 +115,7 @@ def add_result(country: "pycountry.db.Country", points: int) -> None:
 
         sorted_results = [
-            self.get(alpha_2=x[0])
+            cast(pycountry.db.Country, self.get(alpha_2=x[0]))
```
From looking at the code this should not be `Optional` since how the keys are found, but the `self.get(...)` call returns an optional value, so cast this to non optional.

terencehonles commented on src/pycountry/__init__.py:
```diff
@@ -284,21 +302,21 @@ def add_result(
 
 
 # Initialize instances with type hints
-countries: ExistingCountries = ExistingCountries(
```
The typing isn't needed since it's defined from the right hand side of the assignment

terencehonles commented on src/pycountry/db.py:
```diff
@@ -62,31 +64,28 @@ class Subdivision(Data):
 
 
-def lazy_load(f):
+def lazy_load(f: F) -> F:
```
Let mypy or other type checkers know that they should consider whatever to be returned to match whatever was passed in. I can also call this `T` since it's doesn't have a lower bound.

terencehonles commented on src/pycountry/db.py:
```diff
@@ -62,31 +64,28 @@ class Subdivision(Data):
-    data_class: Union[Type, str]
-    root_key: Optional[str] = None
+class Database(Generic[D]):
+    factory: Type[D]
+    indices: Dict[str, Dict[Any, D]]
     no_index: List[str] = []
+    objects: List[D]
+    root_key: Optional[str] = None
```
I defined the type of a few of these up here since they all aren't defined in `__init__` and it makes it clearer what they are.

terencehonles commented on src/pycountry/db.py:
```diff
@@ -62,31 +64,28 @@ class Subdivision(Data):
         self._is_loaded = False
         self._load_lock = threading.Lock()
 
-        if isinstance(self.data_class, str):
```
For mypy's sake all factory functions have been defined as classes now and will be defined at class creation time.

terencehonles commented on src/pycountry/__init__.py:
```diff
@@ -48,13 +50,13 @@ def remove_accents(input_str: str) -> str:
     root_key = "3166-1"
 
-    def search_fuzzy(self, query: str) -> List[Type["ExistingCountries"]]:
+    def search_fuzzy(self, query: str) -> List[pycountry.db.Country]:
```
`Type[class]` means the class itself (good for factories) not an instance of the class. This also should not return itself, but the underlying data.

terencehonles commented on src/pycountry/db.py:
```diff
@@ -166,17 +165,15 @@ def remove_entry(self, **kw):
 
     @lazy_load
-    def __iter__(self) -> Iterator["Database"]:
+    def __iter__(self) -> Iterator[D]:
```
This is not returning a list of itself, but the underlying data.

terencehonles commented on src/pycountry/db.py:
```diff
@@ -194,7 +191,7 @@ def get(self, **kw: Optional[str]) -> Optional[Any]:
 
     @lazy_load
-    def lookup(self, value: str) -> Type:
+    def lookup(self, value: str) -> D:
```
This should have been `Any` not `Type` since this is not returning a factory, but is returning an instance of `Type`. Now that this is a generic class it returns the generic type.

nschimmoller commented:
@terencehonles the CI failed for the pre-commit linting. Could you please run that locally and push an updated commit to address?

terencehonles commented:
> @terencehonles the CI failed for the pre-commit linting. Could you please run that locally and push an updated commit to address?

Done, I don't have pre-commit on this computer yet, but that should address the lint errors

terencehonles commented:
Sorry, thought I had everything, thanks for that

nschimmoller commented on src/pycountry/tests/test_general.py:
```diff
@@ -317,7 +317,7 @@ def test_add_entry():
     country = pycountry.countries.get(alpha_2="XK")
-    assert isinstance(country, pycountry.countries.data_class)
+    assert isinstance(country, pycountry.countries.factory)
 
```
@terencehonles additionally I'd like to propose amending the `test_general.py` file with the following tests as part of this merge. While I know `mypy` handles type checking in future iterations I'd love to avoid these classes getting changed to something other than expected / anticipated. 

Would be interested in your thoughts as well on this.

``` python
def test_database_instance_of_language():
    assert isinstance(pycountry.languages, pycountry.Languages)


def test_database_instance_of_country():
    assert isinstance(pycountry.countries, pycountry.ExistingCountries)


def test_database_instance_of_subdivision():
    assert isinstance(pycountry.subdivisions, pycountry.Subdivisions)


def test_database_instance_of_script():
    assert isinstance(pycountry.scripts, pycountry.Scripts)


def test_instance_of_currency():
    assert isinstance(list(pycountry.currencies)[0], pycountry.Currency)


def test_instance_of_language():
    assert isinstance(list(pycountry.languages)[0], pycountry.Language)


def test_instance_of_country():
    assert isinstance(list(pycountry.countries)[0], pycountry.Country)


def test_instance_of_script():
    assert isinstance(list(pycountry.scripts)[0], pycountry.Script)


def test_instance_of_currency():
    assert isinstance(list(pycountry.currencies)[0], pycountry.Currency)
```

terencehonles replied:
It looks like a number of these already existed. I used `pytest.mark.parameterize` to collapse these somewhat duplicative tests and added the `isinstance` check on the first item as you suggested.

terencehonles replied:
> While I know `mypy` handles type checking in future iterations I'd love to avoid these classes getting changed to something other than expected / anticipated.

Is mypy configured to run as a check? I'm not 100% sure it is, and that would be a good thing to enable, but I didn't spend ~too much~ any time making sure it does.

nschimmoller replied:
@terencehonles love the `pytest.mark.parameterize` use to collapse these test. In terms of the `mypy` check it isn't included in the Makefile currently but #183 should address that. It is part of the .github/workflows on main under the mypy-check.yml file.

terencehonles replied:
cool

nschimmoller commented on src/pycountry/__init__.py:
```diff
@@ -186,12 +204,12 @@ def parent(self):
     # the country!
 
-    data_class = SubdivisionHierarchy
+    factory = SubdivisionHierarchy
```
@terencehonles thanks so much for this PR. I was pulling my hair out trying to figure out how to get the records within a database class to return their respective type. 

The one record type that I'm not sure makes sense is the class for records in Subdivisions. Currently a record like `Alabama` returns a class type of `SubdivisionHierarchy`. Which doesn't follow the rest of the naming conventions where the Database Container Class (i.e. languages) is named for the plural version of the Data Instance Class (i.e. language) it contains. 

What do you think about this implementation to make this more intuitive to end users?

```
class SubdivisionHierarchy(pycountry.db.Data):
    def __init__(self, **kw):
        if "parent" in kw:
            kw["parent_code"] = kw["parent"]
        else:
            kw["parent_code"] = None
        super().__init__(**kw)
        self.country_code = self.code.split("-")[0]
        if self.parent_code is not None:
            # Split the parent_code to check if the country_code is already present
            parts = self.parent_code.split("-")
            if parts[0] != self.country_code:
                self.parent_code = f"{self.country_code}-{self.parent_code}"

    @property
    def country(self):
        return countries.get(alpha_2=self.country_code)

    @property
    def parent(self):
        if not self.parent_code:
            return None
        return subdivisions.get(code=self.parent_code)


class Subdivision(SubdivisionHierarchy):
    pass


class Subdivisions(pycountry.db.Database[Subdivision]):
    # Note: subdivisions can be hierarchical to other subdivisions. The
    # parent_code attribute is related to other subdivisions, *not*
    # the country!

    factory = Subdivision
```

terencehonles replied:
I'll have to think about this a little. I realize now that I re-exported `Subdivision` but it's not actually used(?) (and I just flagged this for myself adding the tests that @nschimmoller requested). You're suggesting `Subdivision` be a subclass of `SubdivisionHierarchy`, but should we just get rid of the class? The name isn't super important, and I will need to look at the type definition for `Subdivision.search_fuzy` since I believe it may not be correct (like some of the others).

nschimmoller replied:
@terencehonles I think you're correct that the `Subdivision` class is not currently used. 

#### Context
--------------
I tried to add the same type hinting you're adding for instances of the containers. I believe prior to this commit the only instance that would properly return their own Class name was `Country`. In an attempt to reverse engineer how `Country` instances were properly storing the intended name I tried to mimic building a `Subdivision` class in `pycountry.db`. As you can see in #175 then Christian, then cleaned up some of the code I was working on to try and add this functionality. 

Since all tests passed, I incorrectly assumed this was implemented correctly. I think again a learning experience as all of us as maintainers are learning to work together. It seems as though incorrectly implementing this is a great example of Cunningham's Law 🤣

So I really do appreciate your willingness to help contribute to `pycountry`! 

#### Naming
--------------
> should we just get rid of the [Subdivision] class?

It does seem that having both `SubdivisionHierarchy` and `Subdivision` may be redundant. I think I would be open to the idea of removing the existing `Subdivision` class and then renaming `SubdivisionHierarchy` to `Subdivision` to achieve the same desired outcome where the instance of the container is the singular version of the plural container. 

> The name isn't super important

While yes, the name isn't super important, I do think a standard naming convention helps end users from a ease of use. For example having `pycountry.countries.search_fuzzy()` and `pycountry.subdivisions.fuzzy_search()` would be confusing for end users. I too so think breaking the naming convention for instances of containers could be confusing. 

> Need to look at the type definition of Subdivision.search_fuzzy()

Yes currently hinted as a `list` containing type `Subdivisions` it actually is a list containing type `SubdivisionHierarchy`. Type hinting for this will depend upon decisions on prior two points

nschimmoller commented:
> Sorry, thought I had everything, thanks for that

no problem! didn't think a single blank line was worth asking you again. especially if you were spot checking without it installed. 😅

terencehonles commented:
> > Sorry, thought I had everything, thanks for that
> 
> no problem! didn't think a single blank line was worth asking you again. especially if you were spot checking without it installed. 😅

Got it installed now :sweat_smile:, wasn't sure if the project was using black and it was straightforward enough to get it installed (it's a new computer that I'm still breaking in).

terencehonles commented on src/pycountry/tests/test_general.py:
```diff
@@ -17,9 +18,32 @@ def logging():
+
 def test_country_list():
     assert len(pycountry.countries) == 249
-    assert isinstance(list(pycountry.countries)[0], pycountry.db.Data)
```
These are redundant and are wider than the updated `test_expected_instance_types` so I'm letting that test handle the type checking and these handle the counts or other attribute checks

nschimmoller commented on src/pycountry/__init__.py:
```diff
@@ -186,12 +204,12 @@ def parent(self):
+    factory = SubdivisionHierarchy
     no_index = ["name", "parent_code", "parent", "type"]
     root_key = "3166-2"
 
```
This comment is actually in reference to line 265. Type hinting for `Subdivisions.search_fuzzy()` needs updating per @terencehonles's comments.

Adding this as a change request before merging for tracking purposes.

terencehonles replied:
I've updated this and made some changes so there should be a complete review on this since I made modifications based on errors mypy was pointing out.

terencehonles commented on src/pycountry/__init__.py:
```diff
@@ -72,20 +75,18 @@ def add_result(country: "pycountry.db.Country", points: int) -> None:
-            print(candidate)
-            add_result(candidate.country, 49)
+        match_subdivions = subdivisions.match(query=query)
+        for subdivision in match_subdivions:
```
mypy didn't like `candidate` being re-used for different variables types so I renamed these to use separate names per type.

terencehonles commented on src/pycountry/__init__.py:
```diff
@@ -126,101 +124,125 @@ class HistoricCountries(ExistingCountries):
 
 
 class SubdivisionHierarchy(pycountry.db.Data):
-    def __init__(self, **kw):
-        if "parent" in kw:
-            kw["parent_code"] = kw["parent"]
-        else:
-            kw["parent_code"] = None
-        super().__init__(**kw)
-        self.country_code = self.code.split("-")[0]
-        if self.parent_code is not None:
-            # Split the parent_code to check if the country_code is already present
-            parts = self.parent_code.split("-")
-            if parts[0] != self.country_code:
-                self.parent_code = f"{self.country_code}-{self.parent_code}"
```
Instead of performing this logic before and after the parent's `__init__` call, I made these cached properties so they can be deferred, typed correctly, and still would only perform this logic only once.

terencehonles commented on src/pycountry/__init__.py:
```diff
@@ -126,101 +124,125 @@ class HistoricCountries(ExistingCountries):
-    def country(self):
-        return countries.get(alpha_2=self.country_code)
+    def country(self) -> pycountry.db.Country:
+        return cast(
+            pycountry.db.Country, countries.get(alpha_2=self.country_code)
+        )
```
This can also be typed:

```suggestion
        result = countries.get(alpha_2=self.country_code)
        assert result
        return result
```

terencehonles commented on src/pycountry/__init__.py:
```diff
@@ -126,101 +124,125 @@ class HistoricCountries(ExistingCountries):
-class Subdivisions(pycountry.db.Database):
+class Subdivisions(pycountry.db.Database[SubdivisionHierarchy]):
     # Note: subdivisions can be hierarchical to other subdivisions. The
     # parent_code attribute is related to other subdivisions, *not*
     # the country!
 
-    data_class = SubdivisionHierarchy
+    factory = SubdivisionHierarchy
     no_index = ["name", "parent_code", "parent", "type"]
     root_key = "3166-2"
 
-    def _load(self, *args, **kw):
-        super()._load(*args, **kw)
+    def _load(self) -> None:
+        super()._load()
 
         # Add index for the country code.
         self.indices["country_code"] = {}
         for subdivision in self:
-            divs = self.indices["country_code"].setdefault(
-                subdivision.country_code.lower(), set()
+            divs = cast(
+                Set[SubdivisionHierarchy],
+                self.indices["country_code"].setdefault(
+                    subdivision.country_code.lower(), set()  # type: ignore[arg-type]
+                ),
             )
             divs.add(subdivision)
```
Is this actually correct? I typed the indices as `Dict[str, Dict[str, D]]`, but should they be instead typed as `Dict[str, Dict[str, Any]]`? I only quickly ran through the code and I'm not completely sure how or if this index is used. Widening the type would remove the need for the type ignore & cast, but then would allow much wider index types. Alternatively, this index could be declared separately so it could be typed more specifically.

terencehonles commented on src/pycountry/__init__.py:
```diff
@@ -126,101 +124,125 @@ class HistoricCountries(ExistingCountries):
+        if result is default and "country_code" in kw:
             # This handles the case where we know about a country but there
             # are no subdivisions: we return an empty list in this case
             # (sticking to the expected type here) instead of None.
             if countries.get(alpha_2=kw["country_code"]) is not None:
-                return []
-        return subdivisions
+                return []  # type: ignore[return-value]
```
I'm a little confused why this is an empty list since `super().get(...)` will return an `Optional[SubdivisionHierarchy]` and this would have to be typed `| list[]`. For now I ignored the return type, but if this is correct the method signature should be updated.

terencehonles reviewed (commented):
These are some things I figured I'd point out about the changes I was making and some questions I had.

federicobond commented:
Together with @zware we've managed to land many of the changes proposed in this pull request 🎉 

I’d like to acknowledge and thank everyone who contributed to the discussion here. I encourage you to try out the latest changes and open separate issues or pull requests for any remaining improvements you’re interested in.

terencehonles commented:
@federicobond I updated this PR to pull in the most recent changes from `main`, but it looks like there are still a number of issues with the typing. I only quickly glanced at it, but I believe my initial understanding of the types returned is correct and that means there are still some issues.

terencehonles commented:
@nschimmoller It looks like I left this to stagnate, but did you still think we should rename `SubdivisionHierarchy` to `Subdivision`? Does there need to be any discussion of this elsewhere or should that just be done in a follow up PR to not block getting some of these fixes in.

terencehonles commented:
> @nschimmoller It looks like I left this to stagnate, but did you still think we should rename `SubdivisionHierarchy` to `Subdivision`? Does there need to be any discussion of this elsewhere or should that just be done in a follow up PR to not block getting some of these fixes in.

I went ahead and renamed the class as the previous discussion showed it was unused and it's likely there would be agreement to rename the class. Since `Country`, a trivially simple class, is now the only `Data` subclass defined in `pycountry.db` I moved it to be defined in the top level file like all the other concrete implementations I added.

terencehonles commented:
There really aren't too many changes left at this point and I updated the description to mention them. There _is_ a long history of comments, and I tried to clean those up. The two sets of changes in this PR are:

- fixes to typing
- fixing how some of the code was written (calling class methods as if they are static methods and manually specifying the instance to use)

There really aren't any "style" changes, unless you count variable renaming "style" and that was to not confuse the type-checker. The two "sets" of changes are closely related and don't really benefit from splitting up as the type information will just be wrong if it's not included in the changes.

federicobond commented:
What I meant by “style” changes includes things like the removal of explicit type annotations when instantiating the databases in `__init__.py`, moving the Country class from `pycountry.db`, and other small adjustments that aren’t strictly necessary to improve the typing situation.

My suggestion would be to open a new, minimal pull request focused solely on the Subdivision changes, which I believe are the most valuable part. From there, we can reevaluate the remaining changes and decide how to proceed.
